kinode/packages/kns_indexer/kns_indexer/src/lib.rs

```diff
@@ -75,26 +168,27 @@ fn init(our: Address) {
     // us to quickly verify we have the updated mapping with root hash, but right
     // now it's tricky to recover from missed events.
 
-    let state = State {
-        chain_id: CHAIN_ID,
-        contract_address: KIMAP_ADDRESS.parse::<eth::Address>().unwrap(),
-        nodes: HashMap::new(),
-        names: HashMap::new(),
-        last_block: KIMAP_FIRST_BLOCK,
-    };
+    // Create or load versioned state
 
-    if let Err(e) = main(our, state) {
+    let versioned_state = VersionedState::load_or_create();
+
+    if let Err(e) = main(our, versioned_state) {
         println!("fatal error: {e}");
     }
 }
 
-fn main(our: Address, mut state: State) -> anyhow::Result<()> {
+fn main(our: Address, mut versioned_state: VersionedState) -> anyhow::Result<()> {
+    let mut state = versioned_state.state_mut();
+
     #[cfg(feature = "simulation-mode")]
     add_temp_hardcoded_tlzs(&mut state);
 
+    let contract_address = KIMAP_ADDRESS.parse::<eth::Address>().unwrap();
```

Comment:
There is a reason contract address was a field inside state. If the contract address changes, state must be reset. This was particularly an issue in beta as contract address could change -- extremely unlikely to happen on mainnet, but should still keep it in state especially now that we can perform state migrations in this app.